bluefog/common/nccl_controller.cc

```diff
@@ -807,16 +803,45 @@ void NCCLController::NeighborAllreduce(TensorTableEntry& entry) {
     // 2. Local_rank = 0 do the neighbor all with other machines local_rank=0.
     if (mpi_ctx_.local_rank_ == 0) {
       ncclGroupStart();
-      for (size_t i = 0; i < entry.recv_neighbors->size(); ++i) {
-        int recv_rank = entry.recv_neighbors->at(i);
-        void* recvbuf = (void*)(static_cast<const char*>(entry.output->data()) +
-                                num_elements * i * element_size);
-        NCCLCHECK(ncclRecv(recvbuf, num_elements, GetNCCLDataType(entry.tensor),
-                           recv_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
-      }
-      for (int send_rank : *entry.send_neighbors) {
-        NCCLCHECK(ncclSend(sendbuf, num_elements, GetNCCLDataType(entry.tensor),
-                           send_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+      if (!entry.dst_weighting_enabled) {
+        for (size_t i = 0; i < entry.recv_neighbors->size(); ++i) {
+          int recv_rank = entry.recv_neighbors->at(i);
+          void* recvbuf = (void*)(static_cast<const char*>(entry.output->data()) +
+                                  num_elements * i * element_size);
+          NCCLCHECK(ncclRecv(recvbuf, num_elements, GetNCCLDataType(entry.tensor),
+                             recv_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+        }
+        for (int send_rank : *entry.send_neighbors) {
+          NCCLCHECK(ncclSend(sendbuf, num_elements, GetNCCLDataType(entry.tensor),
+                             send_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+        }
+      } else {
+        cudaStreamSynchronize(nccl_ctx_.stream);
```

Comment:
Add a comment for the reason of adding this Sync for future reference

Reply:
Done.